File: Data Structures & Algorithms/climbing-stairs/submission-3.cpp
class Solution {
public:
    unordered_map<int, int> cache;

    int climbStairs(int n) {
        return dfs(n, 0);
    }
    int dfs(int n, int pos){
        if (cache[pos]) {
            return cache[pos];
        }
        // base case
        if (pos == n) {
            return 1;
        }
        else if (pos > n) {
            return 0;
        }
        else {
            return cache[pos] = dfs(n, pos + 1) + dfs(n, pos + 2);
        }

    }
};
// 1+1=2
// 1+2=3
// 2+3=5
// 3+5=8


// public:
//     int climbStairs(int n) {
//         if (n == 1) {
//             return 1;
//         }
//         if (n == 2) {
//             return 2;
//         }
//         int first = 1;
//         int second = 1;
//         for (int i = 2; i < n; i++) {
//             int sum = first + second;
//             first = second;
//             second = sum;
//         }
//         return first + second;
//     }
// };

File: Data Structures & Algorithms/count-paths/submission-3.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        dp[m - 1][n - 1] = 1

        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                dp[i][j] += dp[i + 1][j] + dp[i][j + 1]
        return dp[0][0]

        # cache = [[-1] * n for _ in range(m)]

        # def dfs(r, c):
        #     if r == (m - 1) and c == (n - 1):
        #         return 1
        #     elif r >= m or c >= n:
        #         return 0
        #     if cache[r][c] == -1:            
        #         cache[r][c] = dfs(r + 1, c) + dfs(r, c + 1)
        #     return cache[r][c]
        
        # return dfs(0, 0)
        

        # n = 2 [0][1]